[#45474] CSP: Allow connect-src for all storages a user knows

wielinde opened the pull request:
https://community.openproject.org/work_packages/45474

wielinde committed:
[#45474] CSP: Allow connect-src for all storages a user knows

https://community.openproject.org/work_packages/45474

cbliard commented on modules/storages/spec/lib/append_content_security_policy_spec.rb:
```diff
@@ -0,0 +1,99 @@
+
+# These specs mainly check that error messages from a sub-service
+# (about unsafe hosts with HTTP protocol) are passed to the main form.
+describe OpenProject::Storages::AppendContentSecurityPolicy do
```
The tests pass and cover the functionnality, but the output is dull and hard to understand as it does not accurately describes its behavior:

```
OpenProject::Storages::AppendContentSecurityPolicy
  on happy path
    behaves like appends hosts to CSP
      secure_header helper function for appending CSP is called with correct storage hosts
  without the storage being active in a project
    behaves like appends hosts to CSP
      secure_header helper function for appending CSP is called with correct storage hosts
  without the storage
    behaves like appends hosts to CSP
      secure_header helper function for appending CSP is called with correct storage hosts
  when current user is admin without being a member of any project
    behaves like appends hosts to CSP
      secure_header helper function for appending CSP is called with correct storage hosts
  without correct permission
    behaves like appends hosts to CSP
      secure_header helper function for appending CSP is called with correct storage hosts
```

It looks like whatever the conditions, CSP are applied.

cbliard replied:
I have modified the test file to get the following description instead:

```
OpenProject::Storages::AppendContentSecurityPolicy
  with a project with an active storage
    when current user is an admin without being a member of any project
      adds CSP directive to allow connection to storage host
    when current user is a member of the project with permission to manage file links
      adds CSP directive to allow connection to storage host
    when current user is a member of the project without permission to manage file links
      does not change CSP directives
    when the project is archived
      does not change CSP directives
  with a project without an active storage
    does not change CSP directives
  with a project without any storages configured
    does not change CSP directives
```

I think it better documents the behavior of the class.

Reading it a second time, I would almost rename `AppendContentSecurityPolicy` as `AppendContentSecurityPolicyHook`. It would give a hint that it is called indirectly by some hook mechanism.

cbliard replied:
Or `AppendStoragesHostsToCSPHook`.

cbliard replied:
Renamed as `AppendStoragesHostsToCspHook` in 4c5885d

wielinde replied:
Loving it. So much better to read!

cbliard commented on modules/storages/lib/open_project/storages/append_content_security_policy.rb:
```diff
@@ -0,0 +1,49 @@
+
+    # secure_headers gem provides this helper method to append to the current content security policy
+    controller = context[:controller]
+    controller.append_content_security_policy_directives({ connect_src: hosts })
```
`hosts.empty?` can be tested to return early instead of calling `append_content_security_policy_directives`.

cbliard replied:
Done in 172f992

wielinde replied:
Great catch!

cbliard reviewed (commented):
It works well.

I'd like to modify the behavior to return early when there are no hosts to add to CSP, and I'd also like to have a more descriptive test.

I'll add a commit with the suggested changes. Feel free to review it and merge if you're happy with it.

cbliard committed:
Make test and documentation more readable

Return earlier if there are no hosts to add to the CSP directives.

More descriptive behavior of AppendContentSecurityPolicy in its rspec
specification.

Reduce line length of method documentation to 80 columns.

Use shared_let to reduce test execution time.

Reduce `let` usage to avoid looking up values being used.

cbliard committed:
Rename AppendContentSecurityPolicy as AppendStoragesHostsToCspHook
